fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/RegistryRemovalChecker.java

```diff
@@ -0,0 +1,227 @@
+	private final Set<RegistryKey<?>> missingKeys;
+	private final Set<Pair<Block, String>> missingBlockStates = new HashSet<>();
+
+	@SuppressWarnings("unchecked, rawtypes")
+	public RegistryRemovalChecker(JsonObject root) {
+		JsonObject json = root.getAsJsonObject("entries");
+		Set<RegistryKey<?>> keys = new HashSet<>();
+
+		for (Map.Entry<String, JsonElement> registry : json.entrySet()) {
+			Identifier registryId = Identifier.tryParse(registry.getKey());
+
+			if (registryId == null) continue;
+
+			RegistryKey<? extends Registry<?>> registryRef = RegistryKey.ofRegistry(registryId);
+
+			for (Map.Entry<String, JsonElement> namespacedEntries : registry.getValue().getAsJsonObject().entrySet()) {
+				for (JsonElement entry : namespacedEntries.getValue().getAsJsonArray()) {
+					Identifier entryId = Identifier.of(namespacedEntries.getKey(), entry.getAsString());
+
+					if (entryId == null) continue;
+
+					keys.add(RegistryKey.of((RegistryKey) registryRef, entryId));
+				}
+			}
+
+			Registry<?> clientRegistry = Registries.REGISTRIES.get(registryId);
+
+			if (clientRegistry != null) keys.removeAll(clientRegistry.getKeys());
+		}
+
+		for (Map.Entry<String, JsonElement> blockNsEntry : root.getAsJsonObject("blockStates").entrySet()) {
+			for (Map.Entry<String, JsonElement> blockEntry : blockNsEntry.getValue().getAsJsonObject().entrySet()) {
+				Identifier id = Identifier.of(blockNsEntry.getKey(), blockEntry.getKey());
+
+				if (id == null || !Registries.BLOCK.containsId(id)) continue;
+
+				Block block = Registries.BLOCK.get(id);
+				Set<Pair<Block, String>> missing = new HashSet<>();
+
+				for (JsonElement state : blockEntry.getValue().getAsJsonArray()) {
+					if (block.getStateManager().getProperty(state.getAsString()) == null) {
+						missing.add(Pair.of(block, state.getAsString()));
+					}
+				}
+
+				if (!missing.isEmpty()) {
+					keys.add(RegistryKey.of(RegistryKeys.BLOCK, id));
+					this.missingBlockStates.addAll(missing);
+				}
+			}
+		}
+
+		this.missingKeys = Collections.unmodifiableSet(keys);
+		this.missingNamespaces = keys.stream().map(RegistryKey::getValue).map(Identifier::getNamespace).collect(Collectors.toUnmodifiableSet());
+	}
+
+	public Set<String> getMissingNamespaces() {
+		return missingNamespaces;
+	}
+
+	public Set<RegistryKey<?>> getMissingKeys() {
+		return missingKeys;
+	}
+
+	public Set<Pair<Block, String>> getMissingBlockStates() {
+		return missingBlockStates;
+	}
+
+	public static JsonObject serializeRegistries() {
+		JsonObject json = new JsonObject();
+		json.addProperty("version", VERSION);
+
+		// Sort the entries first
+		Map<String, Map<String, Set<String>>> entriesMap = new TreeMap<>();
+		Map<String, Map<String, Set<String>>> blockStatesMap = new TreeMap<>();
+
+		for (Registry<?> registry : Registries.REGISTRIES) {
+			if (registry.size() == 0 || !RegistryAttributeHolder.get(registry).hasAttribute(RegistryAttribute.REMOVAL_CHECKED)) continue;
```

Comment:
It should only write when modded, this file should not be rewritten if you load with only Fabric API.